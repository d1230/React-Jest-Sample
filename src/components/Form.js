import { useFormik } from "formik";
import * as Yup from "yup";
import React from "react";
import classes from "../styles.module.css";
export default function Form() {
    React.useEffect(()=>{
        console.log(formikForm.errors)
    })
  const formikForm = useFormik({
    initialValues: { id: "", password: "" },
    validationSchema: Yup.object({
      id: Yup.string()
        .min(2, "Minimum 2 Characters")
        .max(8, "Maximum 2 Characters")
        .required("ID is Required"),
      password: Yup.string()
        .min(8, "Minimum 8 Characters")
        .max(18, "Maximum 18 Characters")
        .required("Password is Required"),
    }),
    onSubmit: (values, { resetForm }) => {
      alert(values);
      console.log(values);
      resetForm({ values: "" });
    },
   
  });

  return (
    <>
      <form className={classes.container} onSubmit={formikForm.handleSubmit}>
        <div className="h1">Login Form</div>
        <div>
          <label>ID:</label>
          <input
          aria-label="ID"
          className={formikForm.errors.id ? 'error' : ''}
            type="text"
            name="id"
            placeholder="type your id..."
            onChange={formikForm.handleChange}
            
            value={formikForm.values.id}
          ></input>
          {formikForm.errors.id && formikForm.touched.id && (
            <p className={classes.error} aria-label="idError" data-testid="id-Error-Message">{formikForm.errors.id}</p>
          )}
        </div>

        <div>
          <label>Password:</label>
          <input
          aria-label="Password"
          className={formikForm.errors.password ? 'error' : ''}
            type="text"
            name="password"
            placeholder="type your password..."
            onChange={formikForm.handleChange}
            value={formikForm.values.password}
          ></input>
          {formikForm.errors.password && formikForm.touched.password && (
            <p className={classes.error} aria-label="passwordError" data-testid="password-Error-Message">{formikForm.errors.password}</p>
          )}
        </div>

        <div>
          <button aria-label="LoginButton" type="submit" >LogIn</button>
        </div>
      </form>
    </>
  );
}
