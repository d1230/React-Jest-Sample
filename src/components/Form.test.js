import "@testing-library/jest-dom";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Form from "./Form.js";
import userEvent from "@testing-library/user-event";
const formTestConstants = {
  npass: {
    id: "f",
    password: "12",
  },
  pass: {
    id: "correct",
    password: "1234567890",
  },
};
//it describe
test("failed case: showing password error", async () => {
  //const handleSubmit = jest.fn();
  render(<Form />);
  const idInput = screen.getByLabelText("ID");
  const passwordInput = screen.getByLabelText("Password");

  userEvent.type(idInput, formTestConstants.npass.id);
  userEvent.type(passwordInput, formTestConstants.npass.password);
  const button = screen.getByLabelText("LoginButton");
  fireEvent.click(button);
  await waitFor(() => {
    const idError = screen.getByLabelText("idError");
    const passwordError = screen.getByLabelText("passwordError");
    //expect(idError).toBeInTheDocument();
    expect(passwordError).toBeInTheDocument();
  });
});
test("failed case: showing id error", async () => {
  //const handleSubmit = jest.fn();
  render(<Form />);
  const idInput = screen.getByLabelText("ID");
  const passwordInput = screen.getByLabelText("Password");
  userEvent.type(idInput, formTestConstants.npass.id);
  userEvent.type(passwordInput, formTestConstants.pass.password);
  const button = screen.getByLabelText("LoginButton");
  fireEvent.click(button);
  await waitFor(() => {
    const idError = screen.getByTestId("id-Error-Message");
    // const passwordError = screen.getByLabelText('passwordError');
    expect(idError).toBeInTheDocument();
  });
});
test("successful case: showing No error", async () => {
  //const handleSubmit = jest.fn();
  render(<Form />);
  const idInput = screen.getByLabelText("ID");
  const passwordInput = screen.getByLabelText("Password");
  userEvent.type(idInput, formTestConstants.pass.id);
  userEvent.type(passwordInput, formTestConstants.pass.password);
  const button = screen.getByLabelText("LoginButton");
  fireEvent.click(button);
  await waitFor(() => {
    //const idError = screen.getByTestId('id-Error-Message');
    //const passwordError = screen.getByLabelText('passwordError');
    //expect(idError).toBeNull();
    expect(screen.queryByTestId("id-Error-Message")).not.toBeInTheDocument();
  });
});

test("renders MyComponent correctly", () => {
  const { asFragment } = render(<Form />);
  expect(asFragment()).toMatchSnapshot();
});
