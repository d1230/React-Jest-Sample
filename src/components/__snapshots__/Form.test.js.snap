// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MyComponent correctly 1`] = `
<DocumentFragment>
  <form
    class="container"
  >
    <div
      class="h1"
    >
      Login Form
    </div>
    <div>
      <label>
        ID:
      </label>
      <input
        aria-label="ID"
        class=""
        name="id"
        placeholder="type your id..."
        type="text"
        value=""
      />
    </div>
    <div>
      <label>
        Password:
      </label>
      <input
        aria-label="Password"
        class=""
        name="password"
        placeholder="type your password..."
        type="text"
        value=""
      />
    </div>
    <div>
      <button
        aria-label="LoginButton"
        type="submit"
      >
        LogIn
      </button>
    </div>
  </form>
</DocumentFragment>
`;
